addon/components/connect.js

```diff
@@ -1,98 +1,113 @@
+    set: () => assert(`Cannot set redux property "${key}".`)
+  });
+}
+
+/**
+  Return an object of attrs passed to this Component.
+
+  `Component.attrs` is an object that can look like this:
+
+    {
+      myAttr: {
+        value: 'myValue'
+      }
+    }
 
-  if (!stateToComputed) {
-    stateToComputed = () => ({});
-  }
+  Ember provides that a `get` will return the value:
 
-  return Component => {
+    this.get('myAttr') === 'myValue'
 
-    return Component.extend({
+  @method getAttrs
+  @return {Object} an object of key, value for each attr
+  @private
+*/
+function getAttrs(context) {
+  const attrKeys = Object.keys(context.attrs || {});
+  return getProperties(context, attrKeys);
+}
 
+export default (stateToComputed, dispatchToActions) => {
+  return Component => {
+    return Component.extend({
       redux: service(),
 
       init() {
         const redux = this.get('redux');
 
-        let props = stateToComputed.call(this, redux.getState(), this.getAttrs());
+        if (stateToComputed) {
+          const getState = () => stateToComputed.call(this, redux.getState(), getAttrs(this));
 
-        Object.keys(props).forEach(name => {
-          defineProperty(this, name, computed(() =>
-            stateToComputed.call(this, redux.getState(), this.getAttrs())[name]
-          ).property().readOnly());
-        });
+          let state = getState();
 
-        if (!isEmpty(Object.keys(props))) {
-          this.unsubscribe = redux.subscribe(() => {
-            this.handleChange();
+          Object.keys(state).forEach(key => {
+            defineProperty(this, key, computedReduxProperty(key, () => state));
           });
-        }
 
-        this.actions = Object.assign({},
-          this.actions, dispatchToActions.call(this, redux.dispatch.bind(redux))
-        );
+          const handleChange = () => {
+            const newState = getState();
+            if (newState === state) return;
 
-        this._super(...arguments);
-      },
+            const propsToNotify = changedKeys(state, newState);
 
-      handleChange() {
-        const redux = this.get('redux');
+            state = newState;
 
-        const props = stateToComputed.call(this, redux.getState(), this.getAttrs());
+            if (propsToNotify.length > 0) {
+              run.join(() => {
+                propsToNotify.forEach(key => this.notifyPropertyChange(key));
+              });
+            }
+          };
 
-        const notifyProperties = Object.keys(props).filter(name => {
-          return this.get(name) !== props[name];
-        });
+          const unsubscribe = redux.subscribe(handleChange);
 
-        if (notifyProperties.length > 0) {
-          run.join(() => {
-            notifyProperties.forEach(name => this.notifyPropertyChange(name));
+          this.on('didUpdateAttrs', handleChange);
+          this.one('willDestroyElement', () => {
+            this.off('didReceiveAttrs', handleChange);
+            unsubscribe();
```

Comment:
What are the pros/cons of not exposing a separate `willDestoryElement` hook like we did previously? I feel like the `this.one` / `this.off` is a bit magical myself (assuming for the moment this isn't vanillaJS stuff here and it is only possible because of ember). Thoughts? How married are you to making this change now? pros/cons to making this change? For example: what if someone using the 2X had an override here (or inside the didUpdateAttrs) do they lose anything? Could this be considered a breaking api change? Was `getAttrs` a public api people could have tapp'd into? Or is that considered an intimate api?